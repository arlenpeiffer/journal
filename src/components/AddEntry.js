import React from 'react';
import { connect } from 'react-redux';

import EntryForm from './EntryForm';
import { addEntry } from '../redux/actions/journal';
import { updateLogs } from '../utils';

const AddEntry = props => {
  const { addEntry, history, logs } = props;

  const handleSubmitEntry = entry => {
    addEntry(entry);
    updateLogs(entry, logs);
    history.push('/view');
  };

  return <EntryForm handleSubmitEntry={handleSubmitEntry} />;
};

const mapStateToProps = state => ({
  logs: state.user.logs
});

const mapDispatchToProps = dispatch => ({
  addEntry: entry => dispatch(addEntry(entry))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddEntry);
