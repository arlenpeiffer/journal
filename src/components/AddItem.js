import React, { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Collapse from '@material-ui/core/Collapse';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Icon from '@material-ui/core/Icon';
import OutlinedInput from '@material-ui/core/OutlinedInput';
import Typography from '@material-ui/core/Typography';

import ButtonPrimary from './ButtonPrimary';
import { checkIfLogContainsValue, handlePressEnter } from '../utils';

const useStyles = makeStyles(theme => ({
  button: {
    margin: 0
  },
  errorText: {
    overflow: 'hidden',
    textOverflow: 'ellipsis'
  },
  formControl: {
    maxWidth: '100%'
  },
  inputInner: {
    paddingTop: '8.5px',
    paddingBottom: '8.5px'
  },
  inputOuter: {
    marginRight: theme.spacing(1)
  },
  label: {
    alignItems: 'center',
    cursor: 'pointer',
    display: 'inline-flex',
    '& > :first-child': {
      marginRight: theme.spacing(0.5)
    }
  },
  row: {
    alignItems: 'center',
    display: 'flex'
  },
  wrapper: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    marginLeft: props => props.leftMargin && theme.spacing(1.5)
  }
}));

const stopPropagation = event => event.stopPropagation();

const AddItem = ({ callback, dataSource, ...props }) => {
  const [isExpanded, setIsExpanded] = useState(false);
  const [value, setValue] = useState('');
  const [error, setError] = useState('');

  const classes = useStyles(props);

  const icon = isExpanded ? 'expand_more' : 'add';

  const handleBlur = () => {
    document.removeEventListener(
      'keydown',
      stopPropagation,
      true // useCapture
    );
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    document.addEventListener(
      'keydown',
      stopPropagation,
      true // useCapture
    );
  };

  const handleKeyPress = event => {
    handlePressEnter(event, handleSubmit);
  };

  const handleSubmit = () => {
    const formattedValue = value.trim();
    const valueIsDuplicate = checkIfLogContainsValue(
      dataSource,
      formattedValue
    );
    const valueIsEmptyString = formattedValue === '';

    const handleValueIsDuplicate = () => {
      setError(`${value} is already on the list.`);
      setValue('');
    };

    const handleValueIsEmptyString = () => {
      setError('Please enter a value.');
      setValue('');
    };

    const handleSubmitValue = () => {
      callback(formattedValue);
      setError('');
      setValue('');
      setIsExpanded(false);
    };

    if (valueIsEmptyString) {
      return handleValueIsEmptyString();
    } else if (valueIsDuplicate) {
      return handleValueIsDuplicate();
    } else {
      handleSubmitValue();
    }
  };

  const toggleExpanded = () => {
    setIsExpanded(!isExpanded);
  };

  return (
    <div className={classes.wrapper}>
      <span className={classes.label} onClick={toggleExpanded}>
        <Icon fontSize="small">{icon}</Icon>
        <Typography variant="button">Add new</Typography>
      </span>
      <Collapse in={isExpanded} timeout={150}>
        <FormControl className={classes.formControl} error={!!error}>
          <div className={classes.row}>
            <OutlinedInput
              classes={{ input: classes.inputInner }}
              className={classes.inputOuter}
              onBlur={handleBlur}
              onChange={handleChange}
              onFocus={handleFocus}
              onKeyPress={handleKeyPress}
              value={value}
              variant="outlined"
            />
            <ButtonPrimary className={classes.button} onClick={handleSubmit}>
              Add
            </ButtonPrimary>
          </div>
          {error && (
            <FormHelperText className={classes.errorText}>
              {error}
            </FormHelperText>
          )}
        </FormControl>
      </Collapse>
    </div>
  );
};

export default AddItem;
