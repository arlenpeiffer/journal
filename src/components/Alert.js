import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Icon from '@material-ui/core/Icon';
import IconButton from '@material-ui/core/IconButton';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import PropTypes from 'prop-types';

const useStyles = makeStyles(theme => ({
  actionIcon: {
    color: 'white',
    marginLeft: theme.spacing(3)
  },
  content: {
    alignItems: 'center',
    display: 'flex'
  },
  error: {
    backgroundColor: theme.palette.error.dark
  },
  info: {
    backgroundColor: theme.palette.primary.light
  },
  messageIcon: {
    fontSize: 20,
    marginRight: theme.spacing(1.5)
  },
  root: {
    display: 'inline-block',
    marginBottom: theme.spacing(2),
    paddingBottom: 0,
    paddingTop: 0
  },
  warning: {
    backgroundColor: theme.palette.warning.main
  }
}));

const Alert = ({ action, actionIcon, message, variant }) => {
  const classes = useStyles();

  const Content = (
    <div className={classes.content}>
      <Icon className={classes.messageIcon}>{variant}</Icon>
      {message}
      <IconButton
        className={classes.actionIcon}
        disabled={!action}
        onClick={action}
      >
        <Icon>{actionIcon || 'close'}</Icon>
      </IconButton>
    </div>
  );

  return (
    <SnackbarContent
      className={`${classes.root} ${classes[variant]}`}
      message={Content}
    />
  );
};

export default Alert;

Alert.propTypes = {
  action: PropTypes.func,
  actionIcon: PropTypes.string,
  message: PropTypes.string.isRequired,
  variant: PropTypes.oneOf(['error', 'info', 'warning']).isRequired
};
