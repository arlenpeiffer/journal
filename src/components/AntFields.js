import React from 'react';
import {
  DatePicker as AntDatePicker,
  Form,
  Input as AntInput,
  InputNumber as AntInputNumber,
  Radio as AntRadio,
  Select as AntSelect,
  Switch as AntSwitch
} from 'antd';

const CreateAntField = Component => ({
  field,
  form,
  // hasFeedback,
  label,
  submitCount,
  type,
  ...props
}) => {
  const error = form.errors[field.name];
  const onBlur = () => form.setFieldTouched(field.name, true);
  // const onChange = value => form.setFieldValue(field.name, value);
  // const onInputChange = event => {
  //   const value = event.target.value;
  //   form.setFieldValue(field.name, value);
  // };
  const submitted = submitCount > 0;
  const submittedError = error && submitted;
  const touched = form.touched[field.name];
  const touchedError = error && touched;
  return (
    <Form.Item
      // hasFeedback
      help={submittedError || touchedError ? error : undefined}
      label={label}
      validateStatus={submittedError || touchedError ? 'warning' : ''}
    >
      <Component
        {...field}
        {...props}
        onBlur={onBlur}
        // onChange={type ? onInputChange : onChange}
      />
    </Form.Item>
  );
};

export const DatePicker = CreateAntField(AntDatePicker);
export const Input = CreateAntField(AntInput);
export const InputNumber = CreateAntField(AntInputNumber);
export const RadioGroup = CreateAntField(AntRadio.Group);
export const Select = CreateAntField(AntSelect);
export const Switch = CreateAntField(AntSwitch);
export const TextArea = CreateAntField(AntInput.TextArea);
