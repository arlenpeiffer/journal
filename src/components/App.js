import React from 'react';
import { connect } from 'react-redux';
import DashboardPage from './DashboardPage';
import LoginPage from './LoginPage';

function App(props) {
  const { loggedIn } = props;
  return <div>{loggedIn ? <DashboardPage /> : <LoginPage />}</div>;
}

const mapStateToProps = state => ({
  loggedIn: state.user.userInfo.loggedIn
});

export default connect(mapStateToProps)(App);
