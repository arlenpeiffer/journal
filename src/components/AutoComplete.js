import React, { useRef } from 'react';
import { useField, useFormikContext } from 'formik';
import { createFilterOptions } from '@material-ui/lab/Autocomplete';
import MuiAutocomplete from '@material-ui/lab/Autocomplete';
import TextField from '@material-ui/core/TextField';

const AutoComplete = ({ dataSource, label, name, ...props }) => {
  const [field, meta] = useField(name);
  const { setFieldValue } = useFormikContext();

  const { error, touched } = meta;
  const hasError = error && touched ? true : false;

  const inputRef = useRef();

  const filterOptions = createFilterOptions({
    matchFrom: 'start',
    trim: 'true'
  });

  const handleBlur = event => {
    field.onBlur(event);
    setFieldValue(field.name, inputRef.current.value);
  };

  const handleChange = (e, value) => {
    setFieldValue(field.name, value);
  };

  const handleInputChange = (e, value) => {
    const valueIsEmptyString = value === '';
    const valueHasLengthOfOne = value.length === 1;

    if (valueIsEmptyString || (valueHasLengthOfOne && hasError)) {
      setFieldValue(field.name, value);
    }
  };

  const Input = params => (
    <TextField
      error={hasError}
      fullWidth
      helperText={hasError && error}
      InputProps={{ ...params.InputProps, type: 'search' }}
      inputRef={inputRef}
      label={label}
      {...params}
    />
  );

  return (
    <MuiAutocomplete
      autoComplete={true}
      disableClearable
      disableOpenOnFocus
      filterOptions={filterOptions}
      freeSolo
      id={name}
      onBlur={handleBlur}
      onChange={handleChange}
      onInputChange={handleInputChange}
      options={dataSource}
      renderInput={Input}
      value={field.value}
      {...props}
    />
  );
};

export default AutoComplete;
