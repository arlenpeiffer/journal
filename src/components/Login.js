import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'antd';
import { startLogin } from '../redux/actions/userInfo';

export function Login(props) {
  const { startLogin } = props;
  return (
    <div>
      <Button onClick={startLogin} type="primary">
        Sign in with Google
      </Button>
    </div>
  );
}

const mapDispatchToProps = dispatch => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(
  undefined,
  mapDispatchToProps
)(Login);
