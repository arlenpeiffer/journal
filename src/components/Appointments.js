import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Field, FieldArray } from 'formik';
import { AutoComplete, Input } from './AntFields';
import { Button, Form, Icon, Popconfirm } from 'antd';

function Appointments(props) {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedField, setSelectedField] = useState(undefined);
  const [selectedIndex, setSelectedIndex] = useState(undefined);
  const { appointments, logs, setFieldValue } = props;

  const handleBlur = () => {
    resetState();
  };

  const handleChange = (value, index, field) => {
    if (!value) {
      resetState();
    } else {
      if (!isOpen) {
        setIsOpen(true);
        setSelectedField(field);
        setSelectedIndex(index);
      }
    }
    setFieldValue(`appointments[${index}].${field}`, value);
  };

  const handleFocus = (index, field) => {
    setIsOpen(false);
    setSelectedField(field);
    setSelectedIndex(index);
  };

  const handleOpen = (index, field) =>
    selectedField === field && selectedIndex === index && isOpen;

  const handleSelect = () => {
    resetState();
  };

  const resetState = () => {
    setIsOpen(false);
    setSelectedField(undefined);
    setSelectedIndex(undefined);
  };

  return (
    <div id="appointments">
      <Form.Item label="Appointments">
        <FieldArray
          name="appointments"
          render={({ push, remove }) => (
            <div>
              {appointments.map((appointment, index) => (
                <Form.Item key={index}>
                  <Field
                    autoComplete="off"
                    component={AutoComplete}
                    dataSource={logs.appointments}
                    filterOption={(inputValue, option) => {
                      const appointment = option.key.toLowerCase();
                      const value = inputValue.toLowerCase();
                      return appointment.startsWith(value);
                    }}
                    name={`appointments[${index}].type`}
                    onBlur={handleBlur}
                    onChange={value => {
                      handleChange(value, index, 'type');
                    }}
                    onFocus={() => handleFocus(index, 'type')}
                    onSelect={handleSelect}
                    open={handleOpen(index, 'type')}
                    placeholder="Type of appointment (ex: Therapy, Accupunture, Rhuematology)"
                    style={{ marginBottom: 0 }}
                  />
                  <Field
                    autoComplete="off"
                    component={AutoComplete}
                    dataSource={logs.practitioners}
                    filterOption={(inputValue, option) => {
                      const practitioner = option.key.toLowerCase();
                      const value = inputValue.toLowerCase();
                      return practitioner.startsWith(value);
                    }}
                    name={`appointments[${index}].practitioner`}
                    onBlur={handleBlur}
                    onChange={value => {
                      handleChange(value, index, 'practitioner');
                    }}
                    onFocus={() => handleFocus(index, 'practitioner')}
                    onSelect={handleSelect}
                    open={handleOpen(index, 'practitioner')}
                    placeholder="Practitioner"
                    style={{ marginBottom: 0 }}
                  />
                  <Field
                    autoComplete="off"
                    component={Input}
                    name={`appointments[${index}].notes`}
                    placeholder="Notes about the appointment"
                    style={{ marginBottom: 0 }}
                  />
                  <Popconfirm
                    cancelText="No"
                    okText="Yes"
                    onConfirm={() => remove(index)}
                    title={'Are you sure you want to delete this appointment?'}
                  >
                    <Button type="primary">Remove</Button>
                  </Popconfirm>
                </Form.Item>
              ))}
              <Button
                ghost={true}
                onClick={() => push({ type: '', practitioner: '', notes: '' })}
                type="primary"
              >
                <Icon type="plus" />
                Add Appointment
              </Button>
            </div>
          )}
        />
      </Form.Item>
    </div>
  );
}

const mapStateToProps = state => ({
  logs: state.user.logs
});

export default connect(mapStateToProps)(Appointments);
