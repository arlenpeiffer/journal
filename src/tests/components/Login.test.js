import React from 'react';
import { shallow } from 'enzyme';
import { Login } from '../../components/Login';

test('should render Login correctly', () => {
  const wrapper = shallow(<Login />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startLogin when button is clicked', () => {
  const startLogin = jest.fn();
  const wrapper = shallow(<Login startLogin={startLogin} />);
  wrapper.find('Button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});
