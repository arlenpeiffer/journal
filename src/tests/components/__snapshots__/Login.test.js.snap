// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login correctly 1`] = `
<div>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    type="primary"
  >
    Sign in with Google
  </Button>
</div>
`;
